WalletWasabi/WabiSabi/Models/Decomposition/PossibleDecompositions.cs

```diff
@@ -0,0 +1,194 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using NBitcoin;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.WabiSabi.Models.Decomposition
+{
+	public class PossibleDecompositions
+	{
+		public PossibleDecompositions(
+			IEnumerable<Money> nominalValues,
+			Money maximumTotalValue,
+			Money minimumTotalValue,
+			int maxOutputs)
+		{
+			// 8 is more than enough to be unreasonable in practically any
```

Comment:
It was 6 somewhere else. Now it's 8 or 6?

Reply:
Consider the values that you can construct from the std denom set, when allowing for combinations up to a certain size, and the gaps between these values. Whenever a value falls in such a gap, it must be rounded down to the nearest sum that is possible to construct.

If add up the sizes of the gaps smaller than 1000 sats or 0.005%, for combinations up to size 3, only about 35% of the values up to 1.0 BTC fall within such a window. Going up to size 4, 99.7% of the values fall within such gaps, and 94% of values fall within 1000 sats. Scaling up to 5, 96% of values up to 100.0 BTC fall within 1000 sats, and there are no 
gaps larger than the dust threshold for values up to 1.0 BTC.

In other words, 6 outputs should be more than enough for the vast majority most real world situations. But there are still edge cases.

However, when dealing with extremely large amounts, we potentially need more. We might also prefer arbitrarily larger decompositions for privacy reasons when the fee cost is negligible, even if it's not size optimal, especially when consolidating a larger number of inputs to do so. However, inflating decompositions is easy, for example choose a random value or two, and then exhaustively decompose the remainder, or split the amount randomly in half and decompose the two halves.

For these reasons a hard limit of 8 makes sense, because I can't think of a good reason ever allow more than that. Even if we see a hard fork splitting sats up to a smaller unit, that would likely be because most people won't have billions of sats. Allowing more is unwise, because the complexity is still exponential in the size of the decomposition.

And concretely, the reason for the limit of 6 in `CoinJoinClient` is just because it still consumes about 5000% more RAM than it actually needs.

The code to reduce that memory consumption is not pretty or easy to review, but if/when that's merged, we can raise this limit of 6 from the coinjoin client when it makes sense (e.g. when the user has a balance larger than 10 BTC, etc).

For now on testnet this dramatically reduces the fragmentation behavior that the current amount decomposer implements. The way the likely denominations was implemented is kind of hacky, and when coinjoining, especially without counterparties, it has a positive feedback loop where every transaction will lower the average value of the outputs compared to the inputs, which in turn lowers the average value of the inputs to the next transactions, so as time progresses the wallet will eventually be reduced to dust.

Reply:
Can you set it as a constant with the value of six and set it as a constant and a default value for maxOutputs?

Reply:
I'm going to add a constant for the hard limit, but I won't set it as the default value for `maxOutputs`, only in the guard, because it never really makes sense not to specify a value based on what the client requirest.